src/utils/havoc.js

```diff
@@ -307,7 +309,7 @@ class ObjectValueHavocingVisitor {
       invariant(record instanceof DeclarativeEnvironmentRecord);
       this.visitDeclarativeEnvironmentRecordBinding(record, remainingHavocedBindings);
 
-      if (record instanceof FunctionEnvironmentRecord) {
+      if (record instanceof FunctionEnvironmentRecord && !sideEffectfulWrapper) {
```

Comment:
So this part is because we want to havoc everything that is captured? That will only effect bindings but not objects that are captured.

So `__safeSideEffect(() => x++)` might work but not `__safeSideEffect(() => obj.x++` since only obj will be havoced, not what's inside it. But if you do it deeply, it'll pull in the whole environment with it including the global scope.

It's also weird that order matters now. Any optimized functions before this will read this value assuming initial value and any optimized functions that runs after this will assume havoced value.